## Stacks batch 7 - nb 1

<br>

**Batch 7** contains *all sequenced samples*, meaning samples from all lanes 1 through 5. This will be the first batch that includes samples from Lane 5. Lane 5 includes (1) 11 remaining samples unsequenced, (2) samples with slightly degraded DNA, and (3) reruns of samples that were possible contaminated during the library prep process. Samples that fit under the (2)nd definition have the name of that sample followed by `_c` to show that this is a sample created by combining two separate DNA extractions.  Samples that fit under the (3)rd definition have the name of that sample followed by `_2` to show that this is the second time the sample is being sequenced, but should be treated as a replacement for the first attempt, rather than a replicate. 


<br>

This notebook includes: 
1. `process_radtags` for Lane 5
2. `ustacks` for Lane 5
3. `cstacks - populations` for Lanes 1 through 5


<br>
<br>

#### 8/14/2017

### process_radtags, Lane 5


The barcode file for the lane took a while to put together, since some of these samples are potential replacements for other samples. In order to avoid writing over existing files, I named the replacements with an additional "`_2`". Since I am a little worried about the samples that I prepared by combining two different extractions, I marked these with a "`_c`".

The process_radtags code used here is the same that was used on Lanes 2 and 3 (other single read lanes) in [this notebook](https://github.com/mfisher5/PCod-Korea-repo/blob/master/notebooks/batch_4/Stacks%20batch%204%20-%201.ipynb). 

In [ ]:
#ran in terminal
process_radtags -f /media/mfisher5/New\ Volume/Mary/Raw\ Data/1275_S13_L008_R1_001.fastq.gz \
-i gzfastq \
-y gzfastq \
-o samplesT142 \
-b scripts/barcodes_L5.txt \
-e sbfI \
-E phred33 \
-r -c -q -t 142 2>> samplesT142/process_radtags_outputL5.txt

`Processing single-end data.
Using Phred+33 encoding for quality scores.
Reads will be truncated to 142bp
Found 1 input file(s).
Searching for single-end, inlined barcodes.
Loaded 70 barcodes (6bp).
Will attempt to recover barcodes with at most 1 mismatches.
Processing file 1 of 1 [1275_S13_L008_R1_001.fastq.gz]
  Processing RAD-Tags...1M...2M...3M...4M...5M...6M...7M...8M...9M...10M...11M...12M...13M...14M...15M...16M...17M...18M...19M...20M...21M...22M...23M...24M...25M...26M...27M...28M...29M...30M...31M...32M...33M...34M...35M...36M...37M...38M...39M...40M...41M...42M...43M...44M...45M...46M...47M...48M...49M...50M...51M...52M...53M...54M...55M...56M...57M...58M...59M...60M...61M...62M...63M...64M...65M...66M...67M...68M...69M...70M...71M...72M...73M...74M...75M...76M...77M...78M...79M...80M...81M...82M...83M...84M...85M...86M...87M...88M...89M...90M...91M...92M...93M...94M...95M...96M...97M...98M...99M...100M...101M...102M...103M...104M...105M...106M...107M...108M...109M...110M...111M...112M...113M...114M...115M...116M...117M...118M...119M...120M...121M...122M...123M...124M...125M...126M...127M...128M...129M...130M...131M...132M...133M...134M...135M...136M...137M...138M...139M...140M...141M...142M...143M...144M...145M...146M...147M...148M...149M...150M...151M...152M...153M...154M...155M...156M...157M...158M...159M...160M...161M...162M...163M...164M...165M...166M...167M...168M...169M...170M...171M...172M...173M...174M...175M...176M...177M...178M...179M...180M...181M...182M...183M...184M...185M...186M...187M...188M...189M...190M...191M...192M...193M...194M...195M...196M...197M...198M...199M...200M...201M...202M...203M...204M...205M...206M...207M...208M...209M...210M...211M...212M...213M...214M...215M...216M...217M...218M...219M...220M...221M...222M...223M...224M...225M...226M...227M...228M...229M...230M...231M...232M...233M...234M...235M...236M...237M...238M...239M...240M...241M...242M...243M...244M...245M...246M...247M...248M...249M...250M...251M...252M...253M...254M...255M...256M...257M...258M...259M...260M...261M...262M...263M...264M...265M...266M...267M...268M...269M...270M...271M...272M...273M...274M...275M...276M...277M...278M...279M...280M...281M...282M...283M...284M...285M...286M...287M...288M...289M...290M...291M...292M...293M...294M...295M...296M...297M...298M...299M...300M...301M...302M...303M...304M...305M...306M...307M...308M...309M...310M...311M...312M...313M...314M...315M...316M...317M...318M...319M...320M...321M...322M...323M...324M...325M...326M...327M...328M...329M...330M...331M...332M...333M...334M...335M...336M...337M...338M...339M...340M...341M...342M...343M...344M...345M...346M...347M...348M...349M...350M...351M...352M...353M...354M...355M...356M...357M...358M...359M...360M...361M...362M...363M...364M...365M...366M...367M...368M...369M...370M...371M...372M...373M...374M...375M...376M...377M...378M...379M...380M...381M...382M...383M...384M...385M...386M...387M...388M...389M...390M...391M...392M...393M...394M...395M...396M...397M...398M...399M...400M...401M...
  401247240 total reads; -35483440 ambiguous barcodes; -60796955 ambiguous RAD-Tags; +35978738 recovered; -837191 low quality reads; 304129654 retained reads.`
  
`Closing files, flushing buffers...`

`Outputing details to log: 'samplesT142/process_radtags.log'`
<br>
<br>
<br>

`401247240 total sequences`

 `35483440 ambiguous barcode drops (8.8%)`
 
  ` 837191 low quality read drops (0.2%)`
   
` 60796955 ambiguous RAD-Tag drops (15.2%)`
 
`304129654 retained reads (75.8%)`

In [17]:
pwd

u'/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo'

In [18]:
cd samplesT142/

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/samplesT142


<br>
<br>
<br>

### ustacks, Lane 5

In [1]:
pwd

u'/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/notebooks'

In [2]:
cd ../

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo


In [3]:
!mkdir stacks_b7

In [3]:
cd scripts

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/scripts


In [4]:
!head ustacks_counts_genshell8-14.py

In [5]:
!python ustacks_counts_genshell8-14.py barcodes_L5.txt

In [6]:
!head ustacks_counts_lane5.sh

#!/bin/bash
cd /mnt/hgfs/Shared\ Drive\ D/Pacific\ cod/DataAnalysis/PCod-Korea-repo/scripts

#ustacks
ustacks -t fastq -f samplesT142/JUK07_10.fq -r -d -o stacks_b7 -i 280 -m 5 -M 3 -p 6 --model_type bounded
ustacks -t fastq -f samplesT142/JUK07_13.fq -r -d -o stacks_b7 -i 281 -m 5 -M 3 -p 6 --model_type bounded
ustacks -t fastq -f samplesT142/JUK07_21.fq -r -d -o stacks_b7 -i 282 -m 5 -M 3 -p 6 --model_type bounded
ustacks -t fastq -f samplesT142/JUK07_23.fq -r -d -o stacks_b7 -i 283 -m 5 -M 3 -p 6 --model_type bounded
ustacks -t fastq -f samplesT142/JUK07_34.fq -r -d -o stacks_b7 -i 284 -m 5 -M 3 -p 6 --model_type bounded
ustacks -t fastq -f samplesT142/JUK07_04.fq -r -d -o stacks_b7 -i 285 -m 5 -M 3 -p 6 --model_type bounded


*notice how I started at an ID of 280 -- this is because in batch 6, I had sample files up to an id of 264 in my ustacks script. however, when I copied over the files from batch 6 ustacks I have 828 files / 3 files per individual = 276 individuals. each individual must have a unique identifier, so to be safe I started with an ID of 280*

In [7]:
!mv ustacks_counts_lane5.sh ../ustacks_counts_lane5.sh

In [8]:
cd ../

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo


Before I can run this, I need to unzip all of my gzipped fastq files.

In [ ]:
#ran at command line

#command one
sample_array=`cat ../scripts/L5_sampleList.txt`

#command two
for file in $sample_array
do
	echo $file
	gzip -d $file.fq.gz
	echo 'file unzipped'
done

In [ ]:
# ran in the command line the shell script from above
./ustacks_counts_lane5.sh

**The counts didn't work, so I used the python script below instead to find the number of loci and number of raw fastq reads for each sample**

In [9]:
pwd

u'/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo'

In [10]:
cd scripts

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/scripts


In [13]:
!head -n 15 countreads_fastq.py

### This script counts the read depth of each process_radtags fastq output file ###


import argparse 

parser = argparse.ArgumentParser(description="count number of consensus seqs in .tags files")

parser.add_argument("-s", "--samples", help="file with list of samples (if paired end, should include .1 at end)")
parser.add_argument("-d", "--directory", help="local path with directory with fastq files")
parser.add_argument("-os", "--outputshell", help="bash shell script file name. must have file extension .sh")
parser.add_argument("-of", "--output", help="text file name to store read depths, with local path. Do not include file extension")

args = parser.parse_args()
import subprocess



In [ ]:
#run in terminal
python countreads_fastq.py \
-s L5_sampleList.txt \
-d ../samplesT142 \
-os countreads_fastq_lane5.sh \
-of lane5_fastq_readcounts

In [1]:
cd ../scripts

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/scripts


In [2]:
!head -n 15 countloci_tagsfiles.py

### This script counts the number of loci in each ustacks tags file ###


import argparse 

parser = argparse.ArgumentParser(description="count number of consensus seqs in .tags files")

parser.add_argument("-s", "--samples", help="file with list of samples")
parser.add_argument("-d", "--directory", help="stacks directory with tags files; can be local path, don't include final '/'")
parser.add_argument("-o", "--output", help="output file name with local path")

args = parser.parse_args()

samplefile= open(args.samples, "r")
outfile = open(args.output, "w")


In [ ]:
#run in terminal
python countloci_tagsfiles.py \
-s L5_sampleList.txt \
-d ../stacks_b7 \
-o locicounts_lane5.txt \



<br>
<br>
#### 8/16/2017

I created an excel spreadsheet which compares the # of reads / # tags per individual, and the graphical comparisons of that data which can be found in [this Jupyter notebook](https://github.com/mfisher5/PCod-Korea-repo/blob/master/notebooks/Stacks%20batch%207%20-%20pipeline%20analyses.ipynb). 

<br>
#### 8/17/2017
I've decided to re-run `cstacks` with a few of my new samples in the catalog. 

<br>
### cstacks --> populations

In [1]:
cd ../scripts

/mnt/hgfs/Pacific cod/DataAnalysis/PCod-Korea-repo/scripts


In [2]:
!head cstacks_populations_genShell_8-17.py

###### Generate Shell Script to Run cstacks --> populations ######

## MF 3/8/2017 modified 4/29/2017
## For US Cod Data modified for Korean Cod Data



## At Command Line: python cstacks_populations_genShell_3-8 ARG1 ARG2
##---- ARG1 = samples for cstacks file 
##---- ARG2 = complete sample list file; in this script should be the population map


In [3]:
!python cstacks_populations_genShell_8-17.py \
samples_for_cstacks_b7.txt \
PopMap_L1-5.txt

In [4]:
!head cstacks_populations_8-17.sh


#cstacks
cstacks -b 7 -s ../stacks_b7/PO010715_28.1 -s ../stacks_b7/PO031715_03 -s ../stacks_b7/PO031715_23 -s ../stacks_b7/PO031715_04 -s ../stacks_b7/PO020515_15.1 -s ../stacks_b7/PO010715_02 -s ../stacks_b7/PO010715_12 -s ../stacks_b7/PO020515_14.1 -s ../stacks_b7/PO031715_19.1 -s ../stacks_b7/PO020515_10.1 -s ../stacks_b7/GE011215_10.1 -s ../stacks_b7/GEO012315_12 -s ../stacks_b7/GE011215_18 -s ../stacks_b7/GE011215_22 -s ../stacks_b7/GEO012315_18 -s ../stacks_b7/GEO012315_02 -s ../stacks_b7/GE011215_20.1 -s ../stacks_b7/GE011215_14.1 -s ../stacks_b7/GE011215_28 -s ../stacks_b7/GEO012315_21 -s ../stacks_b7/NA021015_25 -s ../stacks_b7/NA021015_30 -s ../stacks_b7/NA021015_21.1 -s ../stacks_b7/NA021015_26 -s ../stacks_b7/NA021015_13.1 -s ../stacks_b7/NA021015_23 -s ../stacks_b7/NA021015_16.1 -s ../stacks_b7/NA021015_22.1 -s ../stacks_b7/NA021015_17.1 -s ../stacks_b7/NA021015_14.1 -s ../stacks_b7/YS_121316_13 -s ../stacks_b7/YS121315_16 -s ../stacks_b7/YS_121316_07 -s ../stacks_b7/YS_